### PR TITLE
Person detector from MediaPipe Pose

### PR DESCRIPTION
**Pull request by LaurentBerger**
Ok that's not in the [list](https://github.com/opencv/opencv_zoo/issues/135)

### CODE REVIEW

File: models/pose_estimation_mediapipe/README.md [hunk 1]
```diff
@@ -20,6 +22,23 @@ python demo.py
...
+cmake --build build
+
+# detect on camera input
+./build/opencv_zoo_pose_estimation_mediapipe
+# detect on an image
+./build/opencv_zoo_pose_estimation_mediapipe -m=/path/to/model -i=/path/to/image -v
+# get help messages
+./build/opencv_zoo_pose_estimation_mediapipe -h
+```
+
```
Comment (WanliZhong) [inline models/pose_estimation_mediapipe/README.md:41]: Document should be added in `person_detection_mediapipe` folder. The target name is `opencv_zoo_person_detection_mediapipe`

**Review by WanliZhong — COMMENTED** [review]
👍👍👍Thanks for your contribution! Here are some code need to be modified.

File: models/person_detection_mediapipe/demo.cpp [hunk 2]
```diff
@@ -0,0 +1,2535 @@
...
+    for (int row = 0; row < results.rows; row++)
+    {
+        float score = results.at<float>(row, results.cols - 1);
+        Mat personLandmarks;
+        results.row(row).colRange(4, results.cols - 1).reshape(0, 4).convertTo(personLandmarks, CV_32S);
+
+        Point hipPoint = personLandmarks.row(0);
+        Point fullBody = personLandmarks.row(1);
+        Point shoulderPoint = personLandmarks.row(2);
+        Point upperBody = personLandmarks.row(3);
```
[the remark this answers is not in the document]
Comment (LaurentBerger) [reply models/person_detection_mediapipe/demo.cpp:184]: weird : no error, no warnings using vs2022

[the remark this answers is not in the document]
Comment (WanliZhong) [reply models/person_detection_mediapipe/demo.cpp:184]: I got this error:
```shell
cmake --build build

[ 50%] Building CXX object CMakeFiles/opencv_zoo_person_detection_mediapipe.dir/demo.cpp.o
/Users/wanli/Desktop/OpenCV_China/opencv_zoo/models/person_detection_mediapipe/demo.cpp:181:15: error: no viable conversion from 'cv::Mat' to 'cv::Point' (aka 'Point_<int>')
        Point hipPoint = personLandmarks.row(0);
              ^          ~~~~~~~~~~~~~~~~~~~~~~
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:174:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const cv::Point_<int> &' for 1st argument
    Point_(const Point_& pt) = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:175:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'cv::Point_<int> &&' for 1st argument
    Point_(Point_&& pt) CV_NOEXCEPT = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:177:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Size_<int> &' for 1st argument
    Point_(const Size_<_Tp>& sz);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:178:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Vec<int, 2> &' for 1st argument
    Point_(const Vec<_Tp, 2>& v);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1661:28: note: candidate template ignored: could not match 'vector' against 'Point_'
    template<typename _Tp> operator std::vector<_Tp>() const;
                           ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1662:35: note: candidate template ignored: could not match 'Vec' against 'Point_'
    template<typename _Tp, int n> operator Vec<_Tp, n>() const;
                                  ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1663:42: note: candidate template ignored: could not match 'Matx' against 'Point_'
    template<typename _Tp, int m, int n> operator Matx<_Tp, m, n>() const;
                                         ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1665:45: note: candidate template ignored: could not match 'array' against 'Point_'
    template<typename _Tp, std::size_t _Nm> operator std::array<_Tp, _Nm>() const;
                                            ^
/Users/wanli/Desktop/OpenCV_China/opencv_zoo/models/person_detection_mediapipe/demo.cpp:182:15: error: no viable conversion from 'cv::Mat' to 'cv::Point' (aka 'Point_<int>')
        Point fullBody = personLandmarks.row(1);
              ^          ~~~~~~~~~~~~~~~~~~~~~~
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:174:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const cv::Point_<int> &' for 1st argument
    Point_(const Point_& pt) = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:175:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'cv::Point_<int> &&' for 1st argument
    Point_(Point_&& pt) CV_NOEXCEPT = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:177:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Size_<int> &' for 1st argument
    Point_(const Size_<_Tp>& sz);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:178:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Vec<int, 2> &' for 1st argument
    Point_(const Vec<_Tp, 2>& v);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1661:28: note: candidate template ignored: could not match 'vector' against 'Point_'
    template<typename _Tp> operator std::vector<_Tp>() const;
                           ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1662:35: note: candidate template ignored: could not match 'Vec' against 'Point_'
    template<typename _Tp, int n> operator Vec<_Tp, n>() const;
                                  ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1663:42: note: candidate template ignored: could not match 'Matx' against 'Point_'
    template<typename _Tp, int m, int n> operator Matx<_Tp, m, n>() const;
                                         ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1665:45: note: candidate template ignored: could not match 'array' against 'Point_'
    template<typename _Tp, std::size_t _Nm> operator std::array<_Tp, _Nm>() const;
                                            ^
/Users/wanli/Desktop/OpenCV_China/opencv_zoo/models/person_detection_mediapipe/demo.cpp:183:15: error: no viable conversion from 'cv::Mat' to 'cv::Point' (aka 'Point_<int>')
        Point shoulderPoint = personLandmarks.row(2);
              ^               ~~~~~~~~~~~~~~~~~~~~~~
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:174:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const cv::Point_<int> &' for 1st argument
    Point_(const Point_& pt) = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:175:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'cv::Point_<int> &&' for 1st argument
    Point_(Point_&& pt) CV_NOEXCEPT = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:177:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Size_<int> &' for 1st argument
    Point_(const Size_<_Tp>& sz);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:178:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Vec<int, 2> &' for 1st argument
    Point_(const Vec<_Tp, 2>& v);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1661:28: note: candidate template ignored: could not match 'vector' against 'Point_'
    template<typename _Tp> operator std::vector<_Tp>() const;
                           ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1662:35: note: candidate template ignored: could not match 'Vec' against 'Point_'
    template<typename _Tp, int n> operator Vec<_Tp, n>() const;
                                  ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1663:42: note: candidate template ignored: could not match 'Matx' against 'Point_'
    template<typename _Tp, int m, int n> operator Matx<_Tp, m, n>() const;
                                         ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1665:45: note: candidate template ignored: could not match 'array' against 'Point_'
    template<typename _Tp, std::size_t _Nm> operator std::array<_Tp, _Nm>() const;
                                            ^
/Users/wanli/Desktop/OpenCV_China/opencv_zoo/models/person_detection_mediapipe/demo.cpp:184:15: error: no viable conversion from 'cv::Mat' to 'cv::Point' (aka 'Point_<int>')
        Point upperBody = personLandmarks.row(3);
              ^           ~~~~~~~~~~~~~~~~~~~~~~
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:174:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const cv::Point_<int> &' for 1st argument
    Point_(const Point_& pt) = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:175:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'cv::Point_<int> &&' for 1st argument
    Point_(Point_&& pt) CV_NOEXCEPT = default;
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:177:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Size_<int> &' for 1st argument
    Point_(const Size_<_Tp>& sz);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/types.hpp:178:5: note: candidate constructor not viable: no known conversion from 'cv::Mat' to 'const Vec<int, 2> &' for 1st argument
    Point_(const Vec<_Tp, 2>& v);
    ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1661:28: note: candidate template ignored: could not match 'vector' against 'Point_'
    template<typename _Tp> operator std::vector<_Tp>() const;
                           ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1662:35: note: candidate template ignored: could not match 'Vec' against 'Point_'
    template<typename _Tp, int n> operator Vec<_Tp, n>() const;
                                  ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1663:42: note: candidate template ignored: could not match 'Matx' against 'Point_'
    template<typename _Tp, int m, int n> operator Matx<_Tp, m, n>() const;
                                         ^
/Users/wanli/Desktop/OpenCV_China/opencv_build/release-dev/install/include/opencv4/opencv2/core/mat.hpp:1665:45: note: candidate template ignored: could not match 'array' against 'Point_'
    template<typename _Tp, std::size_t _Nm> operator std::array<_Tp, _Nm>() const;
                                            ^
4 errors generated.
make[2]: *** [CMakeFiles/opencv_zoo_person_detection_mediapipe.dir/demo.cpp.o] Error 1
make[1]: *** [CMakeFiles/opencv_zoo_person_detection_mediapipe.dir/all] Error 2
make: *** [all] Error 2

```

File: models/pose_estimation_mediapipe/README.md (hunk 1 above)
Comment (LaurentBerger) [reply models/pose_estimation_mediapipe/README.md:41]: oops!

**Review by WanliZhong — APPROVED** [review]
LGTM

**Comment by LaurentBerger** [comment]
> LGTM
please take a look at https://forum.opencv.org/t/get-same-results-in-c-and-python-code/13656

**Comment by fengyuentau** [comment]
> please take a look at https://forum.opencv.org/t/get-same-results-in-c-and-python-code/13656

Is it a key factor in this pull request? The only place where `convertTo` is used is converting landmarks to int32 at the final stage.

---

`convertTo` uses `saturate_cast` which follows the rounding rule, but `astype` from numpy just truncates anything behind the floating dot.

**Comment by LaurentBerger** [comment]
In this PR there is only one convertTo but I have finished to translate pose_estimation_mediapipe/demo.py and change are visble in image result. It is like reading same file using videocpature or imread. input  file is same but inference is not equal.
MAy be it is a detail but when you want to check result you look for an error that's does not exist

**Comment by LaurentBerger** [comment]
don't push there is a bug

**Pull request merged**